### PR TITLE
ERC721 components

### PR DESCRIPTION
**Pull request by starknetdev**
## Introduced changes

Implements ERC721 components

## Checklist

- [x] Add erc721 component files
- [x] Added relevant tests
- [x] Performed self-review of the code

### CODE REVIEW

**Comment by starknetdev** [comment]
Basic tests written, some review of event logging in `test_erc721_balance` integration tests needed.

File: token/src/components/token/erc721/erc721_balance.cairo [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        fn safe_transfer_internal(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u128,
+            data: Span<felt252>
+        ) {
+            self.transfer_internal(from, to, token_id);
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_balance.cairo:225]: i don't remind the exact discussion about why we didn't support the 'safe' part in prev impl
ideally it should check that receiver implements IERC721Receiver

File: token/src/components/token/erc721/erc721_approval.cairo [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC721Owner: erc721_owner_comp::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        // Helper function for allowance model
+        fn _get_approved(
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_approval.cairo:179]: i like the _ prefix too xd, but we decided to follow cairo/rust convention so no _ prefix

File: token/src/components/token/erc721/erc721_metadata.cairo [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+///
+/// Interface
+///
+
+#[starknet::interface]
+trait IERC721Metadata<TState> {
+    fn name(self: @TState) -> felt252;
+    fn symbol(self: @TState) -> felt252;
+    fn token_uri(self: @TState, token_id: u128) -> felt252;
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_metadata.cairo:24]: we should use`ByteArray` to handle strings (not sure if its yet supported in Models, but should be soon)

File: token/src/components/token/erc721/erc721_metadata.cairo [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        +IWorldProvider<TContractState>,
+        +Drop<TContractState>,
+    > of IERC721Metadata<ComponentState<TContractState>> {
+        fn name(self: @ComponentState<TContractState>) -> felt252 {
+            self.get_meta().name
+        }
+        fn symbol(self: @ComponentState<TContractState>) -> felt252 {
+            self.get_meta().symbol
+        }
+        fn token_uri(self: @ComponentState<TContractState>, token_id: u128) -> felt252 {
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_metadata.cairo:62]: token_id should be u256 to respect the spec in exposed fn even if we limit it to u128

File: token/src/components/token/erc721/erc721_metadata.cairo [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        +Drop<TContractState>,
+    > of InternalTrait<TContractState> {
+        fn get_meta(self: @ComponentState<TContractState>) -> ERC721MetaModel {
+            get!(self.get_contract().world(), get_contract_address(), (ERC721MetaModel))
+        }
+
+        fn get_uri(self: @ComponentState<TContractState>) -> felt252 {
+            // TODO : concat with id when we have string type
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_metadata.cairo:92]: we should be able to do this easily with ByteArray now ?

File: token/src/presets/erc721/enumerable.cairo [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        ref self: TState, from: ContractAddress, to: ContractAddress, token_id: u256
+    );
+    fn enum_safe_transfer_from(
+        ref self: TState,
+        from: ContractAddress,
+        to: ContractAddress,
+        token_id: u256,
+        data: Span<felt252>
+    );
+}
```
Comment (notV4l) [inline token/src/presets/erc721/enumerable.cairo:31]: and add enumerable methods total_supply etc..

File: token/src/components/token/erc721/erc721_enumerable.cairo [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+
+#[derive(Model, Copy, Drop, Serde)]
+struct ERC721EnumerableTotalModel {
+    #[key]
+    token: ContractAddress,
+    total_supply: u128,
+}
+
+#[derive(Model, Copy, Drop, Serde)]
+struct ERC721EnumerableOwnerTotalModel {
```
Comment (notV4l) [inline token/src/components/token/erc721/erc721_enumerable.cairo:56]: you can re-use ERC721BalanceModel here

Comment (notV4l) [reply token/src/components/token/erc721/erc721_enumerable.cairo:56]: seems you did it, so remove if un-used xd

File: token/src/presets/erc721/enumerable.cairo [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+use dojo::world::IWorldDispatcher;
+
+#[starknet::interface]
+trait IERC721EnumerablePreset<TState> {
+    // IERC721
+    fn name(self: @TState) -> felt252;
+    fn symbol(self: @TState) -> felt252;
+    fn token_uri(ref self: TState, token_id: u256) -> ByteArray;
+    fn owner_of(self: @TState, account: ContractAddress) -> bool;
+    fn get_approved(self: @TState, token_id: u256) -> ContractAddress;
```
Comment (notV4l) [inline token/src/presets/erc721/enumerable.cairo:11]: seems balance_of is missing

File: token/src/components/token/erc721/erc721_metadata.cairo (hunk 5 above)
Comment (rsodre) [reply token/src/components/token/erc721/erc721_metadata.cairo]: yes, please! been anxiously waiting for that...

**Comment by rsodre** [comment]
any chances to get this and #71 merged before the StarkHack on June 13th?

**Comment by starknetdev** [comment]
> any chances to get this and #71 merged before the StarkHack on June 13th?

Both are complete on my end and I'm awaiting any further instructions. 🙂

**Pull request merged**